### PR TITLE
Be much more careful about copying data from the signature header

### PR DESCRIPTION
**Pull request by pmatilai**
Only look for known tags, and ensure correct type and size where known
before copying over. Bump the old arbitrary 16k count limit to 16M limit
though, it's not inconceivable that a package could have that many files.
While at it, ensure none of these tags exist in the main header,
which would confuse us greatly.

This is optimized for backporting ease, upstream can remove redundancies
and further improve checking later.

Fixes: RhBug:1935049, RhBug:1933867, RhBug:1935035, RhBug:1934125, ...

Fixes: CVE-2021-3421, CVE-2021-20271

### CODE REVIEW

**Comment by pmatilai** [comment]
(reporter credits added to commit message)

File: lib/package.c [hunk 1]
```diff
@@ -31,84 +31,71 @@ struct pkgdata_s {
...
-	default:
-	    if (!(td.tag >= HEADER_SIGBASE && td.tag < HEADER_TAGBASE))
-		continue;
-	    break;
-	}
-	if (!headerIsEntry(h, td.tag)) {
-	    switch (td.type) {
-	    case RPM_NULL_TYPE:
-		continue;
+	if (headerGet(sigh, xl->stag, &td, HEADERGET_RAW)) {
```
Comment (DemiMarie) [inline lib/package.c:75]:
```suggestion
	if (headerGet(sigh, xl->stag, &td, HEADERGET_RAW | HEADERGET_MINMEM)) {
```

File: lib/package.c [hunk 2]
```diff
@@ -31,84 +31,71 @@ struct pkgdata_s {
...
+	    if (!headerPut(h, &td, HEADERPUT_DEFAULT))
 		break;
-	    }
-	    (void) headerPut(h, &td, HEADERPUT_DEFAULT);
+	    rpmtdFreeData(&td);
 	}
     }
-    headerFreeIterator(hi);
+    rpmtdFreeData(&td);
+
```
Comment (DemiMarie) [inline lib/package.c:92]:
```suggestion

bad:
```

File: lib/package.c [hunk 3]
```diff
@@ -31,84 +31,71 @@ struct pkgdata_s {
...
-	case RPMSIGTAG_PAYLOADSIZE:
-	    td.tag = RPMTAG_ARCHIVESIZE;
-	    break;
-	case RPMSIGTAG_FILESIGNATURES:
-	    td.tag = RPMTAG_FILESIGNATURES;
+    rpmtdReset(&td);
+    for (xl = xlateTags; xl->stag; xl++) {
+	/* There mustn't be one in the main header */
+	if (headerIsEntry(h, xl->xtag))
 	    break;
```
Comment (DemiMarie) [inline lib/package.c:74]:
```suggestion
    rpmtdReset(&td);
    for (xl = xlateTags; xl->stag; xl++) {
	/* There mustn't be one in the main header */
	if (headerIsEntry(h, xl->xtag))
	    goto bad;
    }
    for (xl = xlateTags; xl->stag; xl++) {
```

**Review by DemiMarie — APPROVED** [review]
Thank you!  I included a couple of minor performance suggestions, but those should not delay merging.

**Comment by pmatilai** [comment]
> I included a couple of minor performance suggestions, but those should not delay merging.

But that's exactly what such things tend to do, as I'm now wondering could there be some quirks code, especially very old versions, that cause it to actually rely on the put-sort cycle. It's extremely unlikely, but there have been stranger things (see commit da3a3a14e757ccd517e2eb2a3f0293ff48b3ff7f)

**Comment by pmatilai** [comment]
Split the optimization part to a separate commit so the first one can be safely picked for old versions.
I merged the MINMEM part to the first commit though - I had some kind of reason to not put it there in the first place (probably just that MINMEM is a bit iffy in the first place), but thinking it again, MINMEM is what the header iteration does so in a way this is closer to the original behavior.

**Comment by pmatilai** [comment]
Mind you, the suggested optimization to avoid multiple sorts totally makes sense and never occurred to me at all (too much staring at how it always did it), so thanks for that! Just that with critical fixes needing backports and all, other enhancements such as performance are best kept apart - I planned to do a PR for that separately sooner or later.

**Pull request merged**